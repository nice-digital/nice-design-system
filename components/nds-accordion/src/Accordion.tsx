import * as React from "react";

export const Accordion = () => {
	return <p>I am an accordion!</p>;
};
