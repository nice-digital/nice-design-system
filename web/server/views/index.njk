{% extends "layouts/_layout.njk" %}
{% set title = "NICE Bootstrap v2" %}

{% macro item(title, href, description) %}
	<div class="col-12 col-md-4 col-lg-3">
		<h4 class="h6">{{ title }}</h4>
		<p>
			{{ description }}
		</p>
		<p>
			<a class="btn btn--secondary" href="{{ href }}">Read about {{ title }}</a>
		</p>
	</div>
{% endmacro %}

{% block body %}

	<div class="container">
		<h1>Welcome to NICE Experience</h1>
		<h2 class="p">
			Your source for creating beautiful, consistent experiences across NICE.
			<a href="/about/">Read more about NICE Experience</a>
		</h2>

		<p>
			<a class="btn btn--primary" href="/about/installation">Install</a>
			<a class="btn btn--secondary" target="_blank" rel="external" href="https://github.com/nhsevidence/NICE.Bootstrap2">View on GitHub</a>
		</p>

		<h3 class="heading heading--separator">Style Guide</h3>

		<p>The foundations and patterns that are to be used for faster product development.</p>

		<div class="grid">

			{{ item("Layout", "/style-guide/layout", "Containers, grid and layouts") }}

			{{ item("Typography", "/style-guide/typography", "Headings, paragraphs, featured text, quotations, lists, links") }}

			{{ item("Buttons", "/style-guide/buttons", "Button text, button blocks") }}

			{{ item("Colour", "/style-guide/colour", "Colour contrast, SASS variables, colour palettes") }}

		</div>

		<div class="panel">
			<h3 class="heading heading--separator">Components</h3>

			<p>Further patterns</p>

			<div class="grid">

				{{ item("Breadcrumbs", "/components/breadcrumbs", "Secondary navigation scheme that reveals the user's location") }}

				{{ item("Tabs", "/components/tabs", "Content is separated into different panes, with only one pane viewable at a time.") }}

				{{ item("Alpha/beta/live", "/components/alpha-beta-live", "Phase tags and banner to denote alpha and beta services") }}

				{{ item("Navigation", "/components/navigation", "Lorem ipsum dolor sit amet, consectetur adipisicing elit.") }}

			</div>
		</div>

		<h3 class="heading heading--separator">SASS</h3>

		<p>We use <a href="http://sass-lang.com/" target="_blank" rel="external">SASS</a> as our <a href="https://github.com/showcases/css-preprocessors" target="_blank" rel="external">CSS pre-processor</a> as it's mature, got a great feature set and <a href="https://github.com/sass/libsass" target="_blank" rel="external">libsass</a> is super quick.</p>

		<div class="grid">

			{{ item("Code standards", "/sass/code-standards", "SASS code standards and code style guide") }}

			{{ item("Libraries", "/sass/libraries", "3rd party SASS libraries") }}

			{{ item("Documentation", "/sass/docs", "Auto-generated SASS code documentation") }}

		</div>

		<h3 class="heading heading--separator">JavaScript</h3>

		<div class="grid">

			{{ item("Documentation", "/javascript/docs", "Auto-generated JS code documentation") }}

		</div>

	</div>

{% endblock %}

