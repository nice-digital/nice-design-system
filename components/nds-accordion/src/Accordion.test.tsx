import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import { renderToString } from "react-dom/server";
import userEvent from "@testing-library/user-event";

import { Accordion } from "./Accordion";

describe("Accordion", () => {
	it("should match snapshot", () => {
		render(
			<Accordion title="Some title">
				<p>test para</p>
			</Accordion>
		);

		expect(screen.getByRole("button").parentElement).toMatchSnapshot();
	});

	it("should render the Accordion as open if rendered from server (mimic server side render)", () => {
		//mimic server side rendering and JS disabled
		const accordionHtml = renderToString(
			<Accordion
				title="Server rendered accordion"
				displayTitleAsHeading={true}
				headingLevel={2}
			>
				<h3>Accordion body heading</h3>
				<p>Accordion body content</p>
			</Accordion>
		);

		const container = document.createElement("div");
		container.innerHTML = accordionHtml;
		document.body.appendChild(container);

		const button = screen.getByRole("button");
		const contentDiv = screen.getByLabelText(/Server rendered accordion/i);

		expect(button).toHaveAttribute("aria-expanded", "false");
		expect(contentDiv).toBeVisible();
		expect(contentDiv).not.toHaveAttribute("hidden");

		// Clean up
		document.body.removeChild(container);
	});

	it("should render the Accordion as closed if JS is enabled", () => {
		const accordion = render(
			<Accordion
				title="Server rendered accordion with JS enabled"
				displayTitleAsHeading={true}
				headingLevel={2}
			>
				<h3>Accordion body heading</h3>
				<p>Accordion body content</p>
			</Accordion>
		);

		const accordionHeadingText = screen.getByText(
			/Server rendered accordion with JS enabled/
		);
		expect(accordionHeadingText).toBeVisible();

		const accordionContentText = screen.getByText(/Accordion body content/);
		expect(accordionContentText).toBeInTheDocument();
		expect(accordionContentText).not.toBeVisible();
	});

	it.each([
		["closed", "Show", undefined, undefined],
		["closed", "Expand", undefined, "Expand"],
		["closed", "Show", false, undefined],
		["closed", "Expand", false, "Expand"],
		["open", "Hide", true, undefined],
		["open", "Collapse", true, "Collapse"]
	])(
		"should render %s accordion with label of %s",
		(_expectedState, expectedLabel, defaultOpen, actualLabel) => {
			render(
				<Accordion
					title="Some title"
					open={defaultOpen}
					showLabel={defaultOpen ? undefined : actualLabel}
					hideLabel={defaultOpen ? actualLabel : undefined}
				>
					<p>Body content</p>
				</Accordion>
			);
			const button = screen.getByRole("button");
			expect(button).toHaveAttribute(
				"aria-expanded",
				defaultOpen ? "true" : "false"
			);
			expect(button).toHaveTextContent(`${expectedLabel} Some title`);
		}
	);

	it.each([
		[2, "heading2"],
		[3, "heading3"],
		[4, "heading4"],
		[5, "heading5"],
		[6, "heading6"]
	])(
		"should show a heading element if one is passed and displayTitleAsHeading is true",
		(heading, headingText) => {
			render(
				<Accordion
					title={headingText}
					displayTitleAsHeading={true}
					headingLevel={heading}
				>
					<p>Body content</p>
				</Accordion>
			);

			expect(screen.getByRole("heading")).toHaveClass("accordion__heading");
			expect(screen.getByRole("heading")).toHaveProperty(
				"tagName",
				`H${heading}`
			);
			expect(screen.getByRole("heading")).toHaveTextContent(headingText);
		}
	);

	it("should not show a heading element if displayTitleAsHeading is false", () => {
		render(
			<Accordion title="Some title" headingLevel={2}>
				<p>Body content</p>
			</Accordion>
		);

		expect(screen.queryByRole("heading")).not.toBeInTheDocument();
	});

	it("should wrap string title in span", () => {
		render(
			<Accordion title="Test">
				<p>Body content</p>
			</Accordion>
		);

		expect(screen.getByText("Test")).toHaveProperty("tagName", "SPAN");
	});

	it("should wrap number title in span", () => {
		render(
			<Accordion title={99}>
				<p>Body content</p>
			</Accordion>
		);

		expect(screen.getByText("99")).toHaveProperty("tagName", "SPAN");
	});

	it("should update label when expanded", async () => {
		const user = userEvent.setup();
		render(
			<Accordion title="Test" open={false}>
				<p>Body content</p>
			</Accordion>
		);

		const button = screen.getByText(
			(_content, element) => element?.textContent === `Show Test`,
			{
				selector: "button"
			}
		);

		user.click(button);

		await waitFor(() => {
			expect(button).toHaveTextContent("Hide Test");
		});
	});

	it("should open when default open prop changed from false to true", async () => {
		const { rerender } = render(
			<Accordion title="Test" open={false}>
				<p>Body content</p>
			</Accordion>
		);

		const button = screen.getByText(
			(_content, element) => element?.textContent === `Show Test`,
			{
				selector: "button"
			}
		);

		rerender(
			<Accordion title="Test" open={true}>
				<p>Body content</p>
			</Accordion>
		);

		expect(screen.getByRole("button")).toHaveAttribute("aria-expanded", "true");

		await waitFor(() => {
			expect(button).toHaveTextContent("Hide Test");
		});
	});

	it("should close when default open prop changed from true to false", async () => {
		const { rerender } = render(
			<Accordion title="Test" open={true}>
				<p>Body content</p>
			</Accordion>
		);

		const button = screen.getByText(
			(_content, element) => element?.textContent === `Hide Test`,
			{
				selector: "button"
			}
		);

		expect(button).toHaveAttribute("aria-expanded", "true");

		rerender(
			<Accordion title="Test" open={false}>
				<p>Body content</p>
			</Accordion>
		);

		expect(button).toHaveAttribute("aria-expanded", "false");

		await waitFor(() => {
			expect(button).toHaveTextContent("Show Test");
		});
	});

	it("should have appropriate data tracking attribute", async () => {
		const user = userEvent.setup();
		render(
			<Accordion title="Test" open={false}>
				<p>Body content</p>
			</Accordion>
		);

		const buttonElement = screen.getByText(
			(_, node) => node?.className === "accordion__summary"
		);

		expect(buttonElement).toHaveAttribute("data-tracking", "Show");

		user.click(buttonElement);

		await waitFor(() => {
			expect(buttonElement).toHaveAttribute("data-tracking", "Hide");
		});
	});

	it("should allow the passing of custom classes through className", () => {
		render(
			<Accordion title="test title" className="custom-class-name">
				test content
			</Accordion>
		);

		const accordionHeader = screen.getByRole("button").parentElement;
		const accordion = accordionHeader?.parentElement;
		expect(accordion).toHaveClass("custom-class-name");
	});

	it("should allow the passing of caution variant", () => {
		render(
			<Accordion
				title="test title"
				variant="caution"
				displayTitleAsHeading={true}
				headingLevel={2}
			>
				test content
			</Accordion>
		);

		const accordionHeader = screen.getByRole("button").parentElement;
		const accordion = accordionHeader?.parentElement;
		expect(accordion).toMatchSnapshot();
	});
});
