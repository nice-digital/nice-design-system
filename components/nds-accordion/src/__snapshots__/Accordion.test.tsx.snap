// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion should match snapshot 1`] = `
<details
  class="accordion__details "
>
  <summary
    class="accordion__summary"
    data-tracking="Show"
  >
    <span
      class="accordion__toggleLabel toggle__label"
    >
      <svg
        aria-hidden="true"
        class="toggle__icon"
        fill="none"
        focusable="false"
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M478.152 174c0 10.091-3.525 18.661-10.576 25.712l-186 186c-7.243 7.243-15.909 10.864-26 10.864-10.283 0-18.853-3.621-25.712-10.864l-186-186C36.621 192.853 33 184.283 33 174c0-10.091 3.621-18.757 10.864-26L65 126.576C72.424 119.525 81.09 116 91 116c10.091 0 18.661 3.525 25.712 10.576L255.576 265.44 394.44 126.576c7.051-7.051 15.621-10.576 25.712-10.576 9.909 0 18.576 3.525 26 10.576L467.576 148c7.051 7.424 10.576 16.091 10.576 26Z"
          fill="currentColor"
        />
      </svg>
      Show
    </span>
     
    <div
      class="accordion__title"
    >
      <h3>
        Some title
      </h3>
    </div>
  </summary>
  <div
    class="accordion__content"
  >
    <p>
      test para
    </p>
  </div>
</details>
`;
